Feature screen two presentation

anarahari opened the pull request:
feat : Feature screen two presentation

- Composable functions to display Character Details Screen starting with Image, Character details, Location, Origin and Episodes

- First Screen : Added LaunchedEffect for api call to get characters

- First Screen - Added navigation to character details on click of list item

-  Custom App Bar : dynamic title

- Dimens.Kt file to display custom sizes of text, paddings

- CharacterModel : Added Location model 

- Passed GetCharacterDetailsUseCase as a parameter to viewmodelprovider

- DataMapper : Added location model mapper

- Lintcheck changes

anarahari committed:
feat : Screen Two Domain

- added GetCharacterDetailsUseCase
- added getCharacterDetails function in character repository

anarahari committed:
feat : Screen Two Domain

- lint checks resolved

anarahari committed:
feat : Screen Two Domain

- CharacterModel updated with episode and origin model fields

anarahari committed:
feat : Screen Two Domain

- repo impl function for getCharacterDetails

anarahari committed:
feat : Second Screen Data Layer

- Added graphql for GetCharacterDetails
- Repo impl for GetCharacterDetails
- Added DI for GetCharacterDetailsUseCase in DataModule
- added extension functions in data mapper for origin model and episode model

anarahari committed:
feat : Second Screen Data Layer

- main.yaml changes to run github actions for all branches

anarahari committed:
feat : Second Screen Data Layer

- ktlint checks

anarahari committed:
feat : Feature screen two presentation

- Added Nav graph, nav controller for presentation layer
- Added Navroutes for both screens for routing purpose

anarahari committed:
feat : Feature screen two presentation

- Composable functions to display Character Details Screen starting with Image, Character details, Location, Origin and Episodes

- First Screen : Added LaunchedEffect for api call to get characters

- First Screen - Added navigation to character details on click of list item

-  Custom App Bar : dynamic title

- Dimens.Kt file to display custom sizes of text, paddings

- CharacterModel : Added Location model

- Passed GetCharacterDetailsUseCase as a parameter to viewmodelprovider

- DataMapper : Added location model mapper

- Lintcheck changes

shubhamsinghmutualmobile commented on common/src/main/graphql/GetCharacterDetails.graphql:
```diff
@@ -0,0 +1,29 @@
+            id
+            dimension
+            created
+        }
```
The design requires you to fetch only the `name` and `dimension` values

shubhamsinghmutualmobile commented on common/src/main/graphql/GetCharacterDetails.graphql:
```diff
@@ -0,0 +1,29 @@
+            created
+            air_date
+            episode
+        }
```
The design requires you to fetch only the `id` (optional) and `name` values

shubhamsinghmutualmobile commented on data/src/main/java/com/compose/data/datamapper/DataMapper.kt:
```diff
@@ -15,3 +19,55 @@ fun getCharactersQueryToCharacterModel(results: GetCharactersQuery.Result?): Cha
+            episodes = mapToEpisodeModel(it.character?.episode),
+        )
+    }
+}
```
Using double bang operator (`!!`) should be avoided whenever possible. Kotlin recommends to embrace nullable types instead. We can also get rid of the `orEmpty` extension here and delegate using this extension to the composable/view itself.

shubhamsinghmutualmobile replied:
Make sure to get rid of `orEmpty` from all other models as well. I might not add the same comment over and over on all models.

shubhamsinghmutualmobile commented on data/src/main/java/com/compose/data/datamapper/DataMapper.kt:
```diff
@@ -15,3 +19,55 @@ fun getCharactersQueryToCharacterModel(results: GetCharactersQuery.Result?): Cha
+            )
+        }
+    }
+    return list
```
Use `episodes?.map` instead. Creating a new list is relatively memory-inefficient.

shubhamsinghmutualmobile commented on data/src/main/java/com/compose/data/module/DataModule.kt:
```diff
@@ -26,4 +27,13 @@ object DataModule {
     }
+
+    @Provides
+    @Singleton
```
Try to be mindful of when to create `singleton` instances. Singletons should only be created when using factory instances is considerably costlier in terms of time and/or memory.

shubhamsinghmutualmobile replied:
A usecase is generally a very small class and hence should not need to be created as a singleton. Also, why do we need a `coroutineDispatcher` inside a usecase? 🤔

shubhamsinghmutualmobile commented on data/src/main/java/com/compose/data/repository/CharacterRepositoryImpl.kt:
```diff
@@ -21,4 +23,14 @@ internal class CharacterRepositoryImpl @Inject constructor(private val apolloCli
+        return try {
+            val response = apolloClient.query(GetCharacterDetailsQuery(id)).execute()
+            val results = getCharacterDetailsQueryToCharacterModel(response.data)
+            results
```
This can be shortened by removing `val results =`

shubhamsinghmutualmobile commented on data/src/main/java/com/compose/data/repository/CharacterRepositoryImpl.kt:
```diff
@@ -21,4 +23,14 @@ internal class CharacterRepositoryImpl @Inject constructor(private val apolloCli
+            results
+        } catch (ex: Exception) {
+            CharacterModel()
+        }
```
If the return type of this function is nullable (i.e. `CharacterModel?`), then why are we passing a model instance from the `catch` block? Wouldn't it be more suitable to pass `null` from it?

shubhamsinghmutualmobile commented on domain/src/main/java/com/compose/domain/mapper/CharacterModel.kt:
```diff
@@ -1,10 +1,36 @@
+    val dimension: String? = null
+)
+
+data class EpisodeModel(
```
Nit: Removing the `Model` suffix would make the classes more readable.

> [!NOTE]
> `Nit` is short for `nitpick`, meaning that implementing this comment is `optional`.

shubhamsinghmutualmobile commented on domain/src/main/java/com/compose/domain/usecase/GetCharacterDetailsUseCase.kt:
```diff
@@ -0,0 +1,25 @@
+
+class GetCharacterDetailsUseCase @Inject constructor(
+    private val characterRepository: CharacterRepository,
+    @IoDispatcher private val coroutineDispatcher: CoroutineDispatcher
```
No need for the dispatcher. The task of selecting which thread/dispatcher to use should be delegated to the presentation layer instead (which in our case would be the `ViewModel` class).

anarahari replied:
So are you saying that there is no need to specify flowOn and pass the dispatcher ?

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/navigation/RMNavGraph.kt:
```diff
@@ -0,0 +1,38 @@
+                stringResource(R.string.character_details),
+                onBackButtonPressed = {
+                    navController.navigateUp()
+                }
```
Nit: This can be shortened down to `onBackButtonPressed = navController::navigateUp`

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,335 @@
+}
+
+@Composable
+fun DisplayData(
```
Try using more descriptive composable/function names. `DisplayData` doesn't tell us a lot about what it does.

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,335 @@
+}
+
+@Composable
+fun CharacterScreenLoadData(
```
Try using more descriptive composable/function names. `CharacterScreenLoadData` doesn't tell us a lot about what it does.

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,335 @@
+}
+
+@Composable
+fun CharacterStatus(characterModel: CharacterModel, modifier: Modifier) {
```
There are some general conventions we should use while creating composables. Like: `Modifier should be the first param of a composable.`

Try using a composable lint like [compose-lints](https://slackhq.github.io/compose-lints/) to help yourself.

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/CharacterDetailsScreen.kt:
```diff
@@ -0,0 +1,335 @@
+fun CharacterSpecies(characterModel: CharacterModel, modifier: Modifier) {
+    Row(
+        modifier = modifier.padding(
+            start = MaterialTheme.dimens.padding20dp,
+            top = MaterialTheme.dimens.padding8dp
```
Creating variables with exact `dp` names is not maintainable in long term. Let's say right now its value is `padding20Dp = 20.dp`, but tomorrow the designer updates it to `24.dp`, so wouldn't having `padding20Dp = 24.dp` be a little misleading?

shubhamsinghmutualmobile replied:
Try using more generic names like `paddingExtraSmall`, `paddingSmall`, `paddngMedium`, `paddingLarge`, `paddingExtraLarge`, etc.

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/FirstScreen.kt:
```diff
@@ -15,29 +16,36 @@ import androidx.compose.material3.CircularProgressIndicator
+    modifier: Modifier = Modifier,
+    onNavigateCharacterDetails: (String) -> Unit
+) {
+    LaunchedEffect(Unit) {
+        viewModel.getCharacters()
+    }
```
Writing network calls inside a Composable should be avoided first of all. I understand since the alternatives are more complex, therefore, to keep things simple for now, let's introduce a flag named `areCharactersFetched` in this composable that persists configuration changes and allows you to ensure that you don't make unnecessary network calls on every configuration change.

You can write the following:
```kotlin
@Composable
fun CharacterListScreen() {
   var areCharactersFetched by rememberSaveable { mutableStateOf(false) }
   
   LaunchedEffect(Unit) {
       if (!areCharactersFetched) {
           // Updating the state before making the network call to avoid the scenario where
           // the configuration is changed before the network call is finished executing
           isDataFetched = true
           viewModel.getCharacters()
       }
   }
   
   // Rest of your code
}
```

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/FirstScreen.kt:
```diff
@@ -53,32 +61,38 @@ fun CharacterListScreen(viewModel: CharacterViewModel, modifier: Modifier = Modi
     ) {
-        items(characterList.size, key = { characterList[it].id }) { index ->
-            CharacterItem(character = characterList[index])
+        items(characterList.size, key = { characterList[it].id!! }) { index ->
```
Using the double bang operator (`!!`) should be avoided wherever possible

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/screens/FirstScreen.kt:
```diff
@@ -53,32 +61,38 @@ fun CharacterListScreen(viewModel: CharacterViewModel, modifier: Modifier = Modi
             .fillMaxWidth()
             .padding(8.dp)
+            .clickable {
+                onNavigateCharacterDetails(character.id ?: "")
```
Write `character.id.orEmpty()` instead

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/ui/theme/Dimens.kt:
```diff
@@ -0,0 +1,36 @@
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.dp
+
+data class Dimens(
+    val padding4dp: Dp,
+    val padding8dp: Dp,
+    val padding12dp: Dp,
+    val padding16dp: Dp,
+    val padding20dp: Dp,
+    val margin4dp: Dp,
+    val margin8dp: Dp,
+    val margin12dp: Dp,
+    val cardWidth: Dp
+)
+
+val RMDimens = Dimens(
+    padding4dp = 4.dp,
+    padding8dp = 8.dp,
+    padding12dp = 12.dp,
+    padding16dp = 16.dp,
+    padding20dp = 20.dp,
+    margin4dp = 4.dp,
+    margin8dp = 8.dp,
+    margin12dp = 12.dp,
+    cardWidth = 100.dp
+)
```
As mentioned before, try using generic names like `small`, `medium`, `large` to make these classes more maintainable in long term.

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/ui/theme/Dimens.kt:
```diff
@@ -0,0 +1,36 @@
+    cardWidth = 100.dp
+)
+
+val LocalDimens = compositionLocalOf<Dimens> { error("No dimens provided") }
```
Why are we not providing the `RMDimens` in this line? Providing values using `CompositionLocalProvider` should only be necessary when instantiating the object requires something that can only be provided from a Composable scope (like Context).

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/ui/theme/Theme.kt:
```diff
@@ -41,6 +42,10 @@ fun RMAppTheme(
         typography = RMTypography,
-        content = content
+        content = {
+            CompositionLocalProvider(LocalDimens provides RMDimens) {
```
As mentioned before, no need for this `CompositionLocalProvider` here.

shubhamsinghmutualmobile commented on presentation/src/main/java/com/compose/presentation/viewmodel/CharacterViewModel.kt:
```diff
@@ -16,16 +17,20 @@ import javax.inject.Inject
-    }
+    private val _characterDetailsState: MutableStateFlow<UiState> = MutableStateFlow(UiState())
+    val characterDetailsState: StateFlow<UiState> = _characterDetailsState.asStateFlow()
+
+//    init {
+//        getCharacters()
+//    }
```
Remove unused code

shubhamsinghmutualmobile commented:
@anarahari This PR was too large to review 😅 Please avoid creating large PRs from now on. Try to limit them to 150-200 LOC for now.

anarahari committed:
feat : Feature screen two presentation

- Removed Model suffix from model class names
- Removed isEmpty from model Character class
- Added CharacterDetailsViewModel for character details screen
- Passed appropriate viewmodel class in ViewModelFactoryProvider.kt
- Changed dimens naming convention
